nixos/modules/services/misc/pufferpanel.nix

```diff
@@ -0,0 +1,401 @@
+      description = lib.mdDoc ''
+        Server listen port.
+      '';
+    };
+
+    panel = {
+      enable = lib.mkOption {
+        type = lib.types.bool;
+        default = true;
+        description = lib.mdDoc ''
+          Whether to enable PufferPanel web interface.
+        '';
+      };
+
+      registrationEnabled = lib.mkOption {
+        type = lib.types.nullOr lib.types.bool;
+        default = null;
+        description = lib.mdDoc ''
+          Whether users can register themselves in the web interface.
+
+          To create users when registration is disabled, run
+          {command}`pufferpanel --workDir /var/lib/pufferpanel user add`.
+
+          This option overrides corresponding setting set via web interface
+          after service restart.
+        '';
+      };
+    };
+
+    daemon = {
+      enable = lib.mkOption {
+        type = lib.types.bool;
+        default = true;
+        description = lib.mdDoc ''
+          Whether to enable PufferPanel daemon.
+        '';
+      };
+
+      sftp = {
+        host = lib.mkOption {
+          type = lib.types.str;
+          default = "";
+          example = "::1";
+          description = lib.mdDoc ''
+            Bind to the specified IPv4 or IPv6 address.
+          '';
+        };
+
+        port = lib.mkOption {
+          type = lib.types.port;
+          default = 5657;
+          description = lib.mdDoc ''
+            Listen port for SFTP server.
+          '';
+        };
+      };
+
+      auth = {
+        url = lib.mkOption {
+          type = lib.types.nullOr lib.types.str;
+          default = null;
+          example = "https://panel.example.com/oauth2/token";
+          description = lib.mdDoc ''
+            Auth server URL.
+          '';
+        };
+
+        clientId = lib.mkOption {
+          type = lib.types.nullOr lib.types.str;
+          default = null;
+          description = lib.mdDoc ''
+            OAuth2 client ID.
+          '';
+        };
+
+        clientSecretFile = lib.mkOption {
+          type = lib.types.nullOr lib.types.path;
+          default = null;
+          description = lib.mdDoc ''
+            Path to OAuth2 client secret file.
+          '';
+        };
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.panel.enable || cfg.daemon.enable;
+        message = ''
+          Running PufferPanel with both panel and daemon disabled is not
+          supported.
+        '';
+      }
+    ];
+
+    systemd.services.pufferpanel = {
+      description = "PufferPanel game management server";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+
+      path = cfg.extraPackages;
+      environment = cfg.environment // compatEnv // options;
+
+      script = ''
+        ${lib.concatLines (lib.mapAttrsToList (name: _: ''
+          read -r ${name} <"$CREDENTIALS_DIRECTORY"/${name}
+          export ${name}
+        '') secrets)}
```

Comment:
Why not use systemd's environment and EnvironmentFile?

Reply:
`daemon.auth` block is necessary to configure PufferPanel as a standalone daemon. Option `clientSecretFile`, as the name implies, is an OAuth2 client secret for communicating with panel. However, PufferPanel accepts a plaintext client secret in an environment variable. It can be configured with `EnvironmentFile`, but it’s a common option so I think it deserves native support from the module.

Note that we can’t read directly from the specified path since the service runs as dynamic user and the file may be owned by root, so we use `LoadCredential` to load it.

Reply:
I mean specifically the bash export section. This can be replaced with EnvironmentFile which also supports LoadCredential and DynamicUser. I think you need to use the %d placeholder but it could also be another one.

Reply:
> However, PufferPanel accepts a plaintext client secret in an environment variable.

I don’t think I understand what you want here. EnvironmentFile can load files with environment variables, but not files that contain actual values.

Assume we have a client secret `supersecretvalue` in `/run/secrets/pufferpanel-client-secret`. We want PufferPanel to run with `PUFFER_DAEMON_AUTH_CLIENTSECRET=supersecretvalue` environment variable. We can’t generate an EnvironmentFile with a secret, since that would end up in Nix store. So we have to load our secret from a file at runtime before PufferPanel starts. If we pass the path it via some `NIX_PUFFER_DAEMON_AUTH_CLIENTSECRET_FILE` variable, we won’t be able to read the file since it is likely owned by root and script runs as a dynamic user.

LoadCredential does exactly what we want here: systemd reads the contents the given path and writes file in CREDENTIALS_DIRECTORY that we can read.

Reply:
> If we pass the path it via some `NIX_PUFFER_DAEMON_AUTH_CLIENTSECRET_FILE` variable, we won’t be able to read the file since it is likely owned by root and script runs as a dynamic user.

EnvironmentFile gets read by systemd itself, so that should work.

Reply:
EnvironmentFile reads the environment variables from a text file, LoadCredentials copies a text file so that a service script can read it.

With EnvironmentFile, user would have to use the following file:
```
PUFFER_DAEMON_AUTH_CLIENTSECRET=supersecret
```

With LoadCredentials, it’s just a plain text file that get’s read into PUFFER_DAEMON_AUTH_CLIENTSECRET environment variable in script.
```
supersecret
```

I think that using EnvironmentFile for `clientSecretFile` would feel really awkward since we’d have to either
- document that the file for `clientSecretFile` has a specific format and should set a particular environment variable,
- or remove `clientSecretFile` option and document that a particular environment variable should be set using `environmentFiles`.

Note that client secret is the only secret that makes sense in a standalone PufferPanel daemon configuration (aside from Docker credentials, but that already can and should be configured using `environmentFiles`).

For example, see https://github.com/tie-infra/nix-config/blob/7f6d7e636b49b86826a0b8b4d047f5659ca55193/systems/kazuma/pufferpanel.nix#L7-L22. Here, `./pufferpanel-client-secret.age` contains an encrypted `supersecret` string.

Reply:
> I think that using EnvironmentFile for `clientSecretFile` would feel really awkward since we’d have to either

Yes, we do that already in many places https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=environmentFile because it is convenient and flexible.

Reply:
Ah, thanks, I think I understand you point now. Basically https://github.com/NixOS/nixpkgs/pull/225274#discussion_r1161173589, except that I kept a few potentially useful options.

If I understand correctly, new NixOS modules should prefer documentation over configuration, so we should avoid defining options. In this case, remove `auth` block and document that client secret can be set via an environment variable in a file (since PufferPanel lacks a good documentation page for configuration, I think an explicit example is of great value).

I think we can also remove all other options except `enable`, `package`, `extraPackages`, `extraGroups`, `environment` and `environmentFile` since everything can be set via environment variables. As you’ve said in https://github.com/NixOS/nixpkgs/pull/225274#discussion_r1161958273, it’s likely that the user would want to run PufferPanel behind reverse proxy with TLS (e.g. nginx or caddy), so we don’t lose much from removing `openFirewall` option. What do you think?

Reply:
Did what I’ve said in the last comment.